docs/proposals/deployment-hooks.md

```diff
@@ -0,0 +1,300 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<!-- TAG RELEASE_LINK, added by the munger automatically -->
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.4/docs/proposals/deployment-hooks.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Deployment Lifecycle Hooks
+
+## Abstract
+
+A proposal for implementing a new feature for the - Deployment - resource which will
+enable users to execute arbitrary commands necessary to complete a deployment.
+
+The deployment lifecycle hook is a materialization of the "process" a deployment requires.
+A hook is a way of reducing the cost of implementing a full custom deployment process -
+instead, at logical points in the flow control of the process is handed off to the control
+of user code. Practically, hooks often involve one way transitions in state, such as a
+forward database migration, removal of old values from a persistent store, or the clearing
+of a state cache. A hook is effectively a synchronous event listener with veto power - it
+may return success or failure, or in some cases need reexectution.
+
+Because deployment processes typically involve coupling between state (assumed elsewhere)
+and code (frequently the target of the deployment), it should be easy for a hook to be
+coupled to a particular version of code, and easy for a deployer to use the code under
+deployment from the hook.
+
+## Use Cases
+
+* As a Rails application developer, I want to perform a Rails database migration during
+  the application deployment.
+* As an application developer, I want to invoke my test suite before the deployment is
+  rollout.
+* As an application developer, I want to invoke a cloud API endpoint to notify it of the
+  presence of new code.
+
+### Use Case: Invoke a Cloud API Endpoint
+
+Consider an application whose deployment should result in a cloud API call being invoked
+to notify it of the newly deployed code.
+
+Kubernetes provides container lifecycle hooks for containers within a Pod. Currently,
+post-start and pre-stop hooks are supported. For deployments, post-start is the most
+relevant. Because these hooks are implemented by the Kubelet, the post-start hook provides
+some unique guarantees:
+
+* The hook is executed synchronously during the pod lifecycle.
```

Comment:
@bgrant0607 PostStart is definitely synchronous. According to the [docs](http://kubernetes.io/docs/user-guide/container-environment/#hook-handler-execution) and my testing